File: strings/letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.go
package lettercombinationsofaphonenumber

var phone map[string]string = map[string]string{
	"2": "abc",
	"3": "def",
	"4": "ghi",
	"5": "jkl",
	"6": "mno",
	"7": "pqrs",
	"8": "tuv",
	"9": "wxyz",
}

var ans []string = []string{}

func letterCombinations(digits string) []string {

	ans = []string{}

	if len(digits) == 0 {
		return ans
	}

	backtrack(digits, "")

	return ans
}

func backtrack(digits string, combination string) {
	if len(digits) == 0 {
		ans = append(ans, combination)
	} else {

		letters := phone[digits[:1]]

		for i := 0; i < len(letters); i++ {
			letter := letters[i : i+1]
			backtrack(digits[1:], combination+letter)
		}
	}

}

File: strings/letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.ts
const phone: Record<string, string> = {
  2: 'abc',
  3: 'def',
  4: 'ghi',
  5: 'jkl',
  6: 'mno',
  7: 'pqrs',
  8: 'tuv',
  9: 'wxyz',
};

function letterCombinations(digits: string): string[] {
  const ans: string[] = [];

  if (!digits.length) {
    return ans;
  }

  backtrack(digits, '', ans);

  return ans;
}

function backtrack(digits: string, combination: string, ans: string[]): void {
  if (!digits.length) {
    ans.push(combination);
    return;
  }

  const letters: string = phone[digits[0]];
  for (let i = 0; i < letters.length; i++) {
    const letter: string = letters[i];

    backtrack(digits.substring(1), combination + letter, ans);
  }
}
